### PR TITLE
Focus back on document body after dismissing banners

### LINKED ISSUE #4942: [A11y] focus loss when dismissing install banner
## Short description of your issue:

Screenreader focus on Android becomes lost after dismissing the AMP install banner.

## How do we reproduce the issue?

Please post a public URL and ideally post a reduced test case. E.g. on jsbin.com that exhibits only your issue and nothing else.

1. Activate TalkBack (settings > accessibility > TalkBack)
2. open the install banner from by clicking the two link.
3. Dismiss the banner by doubletapping the dismiss button.
4. Swipe left or right.

Actual result: TalkBack plays the the sound that indicates it has reached the beginning/end of the swipe order, thus it has nowhere to go. Focus can only be regained by exploring to another part of the page by touch.

Expected result: focus should be restored to the page content after the banner is dismissed.

## What browsers are affected?

Anroid (Chrome)

## Which AMP version is affected?

Powered by AMP ⚡ HTML – Version 1473296795712

### PR DESCRIPTION
**Pull request by mkhatib**
Fixes #4942

### CODE REVIEW

File: extensions/amp-app-banner/0.1/amp-app-banner.js [hunk 1]
```diff
@@ -397,6 +397,8 @@ export class AmpAndroidAppBanner extends AbstractAppBanner {
  */
 function handleDismiss(state) {
   hideBanner(state);
+  // Restore focus to document after dismiss for screen readers.
+  state.element.ownerDocument.body./*REVIEW*/focus();
```
Comment (aghassemi) [inline extensions/amp-app-banner/0.1/amp-app-banner.js]: can we put this logic in `removeFromFixedLayer`? I suspect this is an a11y issue for all fixed layer items since fixedLayer is a sibling of body and focus needs to be returned to the actual body.

Comment (mkhatib) [inline extensions/amp-app-banner/0.1/amp-app-banner.js]: The focus is lost because of user interaction with the dismiss button and changing the focus to that button not the removal of the banner itself from the fixed layer. So it wouldn't make sense to focus on the document when we remove something from fixed layer programmatically it might actually break the user current focus.

Comment (aghassemi) [inline extensions/amp-app-banner/0.1/amp-app-banner.js]: true we don't want that in all cases, but I think `removeFromFixedLayer` can still handle this situation so we don't run into this bug again. I think this issue also exists for sidebar, user-notification and sticky-ad IF they are closed/dismissed while focus is inside them, which is not unlikely. 

What do you think about this:
inside `removeFromFixedLayer`, if `document.activeElement` is a child of `element`, then move the focus on body.

Comment (mkhatib) [inline extensions/amp-app-banner/0.1/amp-app-banner.js]: Actually thinking about this some more, this would only address it when we are actually using fixed layer which is currently only kicks off on iOS. @dvoytenko to confirm my assumption. What do you think?

Comment (mkhatib) [inline extensions/amp-app-banner/0.1/amp-app-banner.js]: Talked to Dima, my assumption was wrong, I think your suggestion of checking `document.activeElement` might work. Will test it out.

Comment (aghassemi) [inline extensions/amp-app-banner/0.1/amp-app-banner.js]: Without fixedLayer, there shouldn't be a need to do anything with focus. Removing an element that has the focus from DOM, depending on the browser, should either keep the virtual cursor in the place where the removed DOM element was or move it to top of body as far as I know. (speculation though, worth testing).

Comment (aghassemi) [inline extensions/amp-app-banner/0.1/amp-app-banner.js]: If we find that this problem is not related to fixedLayer, I suggest we do what I described above in `removeElement` function instead of `removeFromFixedLayer`

Comment (mkhatib) [inline extensions/amp-app-banner/0.1/amp-app-banner.js]: Yeah I think `removeElement` is a better place for this, this is not limited to fixed layer fix, it's a known accessibility issue for close buttons on dialogs and modals implementations.

Comment (aghassemi) [inline extensions/amp-app-banner/0.1/amp-app-banner.js]: interesting, so focus gets completely messed up if an element with focus is removed from DOM, ugh..

Comment (mkhatib) [inline extensions/amp-app-banner/0.1/amp-app-banner.js]: Done! Moved focus logic to `dom.removeElement`.

File: src/dom.js [hunk 2]
```diff
@@ -112,6 +112,11 @@ export function documentContainsPolyfillInternal_(doc, element) {
  */
 export function removeElement(element) {
   if (element.parentElement) {
+    const activeElement = element.ownerDocument.activeElement;
+    // Restore focus to document if the focused element is inside the removed one.
+    if (element == activeElement || element.contains(activeElement)) {
```
Comment (jridgewell) [inline src/dom.js]: This'll be expensive, how about a boolean parameter to do this?

**Comment by mkhatib** [comment]
@aghassemi 👀 PTAL

File: src/dom.js (hunk 2 above)
Comment (aghassemi) [inline src/dom.js]: I don't think most caller can know whether to set that boolean or not. That requires them knowing for sure the focus is not in the element they are removing, but focus is controlled by the user.

In terms of expensiveness, I have no idea how much this would effect the overall performance. My guess is negligible but let's get @dvoytenko and @cramforce thoughts here.

File: src/dom.js [hunk 3]
```diff
@@ -112,6 +112,11 @@ export function documentContainsPolyfillInternal_(doc, element) {
  */
 export function removeElement(element) {
   if (element.parentElement) {
+    const activeElement = element.ownerDocument.activeElement;
+    // Restore focus to document if the focused element is inside the removed one.
```
Comment (aghassemi) [inline src/dom.js]: nit: maybe just mention accessibility as well here in case someone is wondering why it matters to move the focus to body or not.

**Comment by aghassemi** [comment]
I am happy with this, but let's see what others think of performance implications.

File: src/dom.js (hunk 3 above)
Comment (mkhatib) [inline src/dom.js]: Done

**Comment by aghassemi** [comment]
Testing a simple [example](https://rawgit.com/aghassemi/27c9a5dc37a9112b7e9f08dff7962f91/raw/4b5abce269b6effd7fab17fe737bb634308c95e5/gistfile1.html) with TalkBack. I don't see a problem with focus getting messed up. Upon removing a node that has focus, TalkBack's virtual cursor seems to move to top of body and swipes start reading from the beginning of body. I also tested a sticky-ad amp page with TalkBack and dismissing the sticky ad, did the same. There is something else going on here, I don't think we need this fix.

**Comment by mkhatib** [comment]
Have you tested it with `amp-app-banner`? Is the problem then only happening with the `amp-app-banner`?

File: src/dom.js [hunk 4]
```diff
@@ -112,6 +112,12 @@ export function documentContainsPolyfillInternal_(doc, element) {
  */
 export function removeElement(element) {
   if (element.parentElement) {
+    const activeElement = element.ownerDocument.activeElement;
+    // For accessibility reasons, restore focus to document if the focused
```
Comment (cramforce) [inline src/dom.js]: What do browsers do by default? If this is good for a11y than then is it not the default?

File: src/dom.js [hunk 5]
```diff
@@ -112,6 +112,12 @@ export function documentContainsPolyfillInternal_(doc, element) {
...
+    // element is inside the removed one.
+    if (element == activeElement || element.contains(activeElement)) {
+      element.ownerDocument.defaultView./*REVIEW*/focus();
```
Comment (cramforce) [inline src/dom.js]: `.focus()` calls need to be wrapped in a try block. At least in IE they can fail.

Probably good to make a `focus(element)` function that does that.

**Comment by aghassemi** [comment]
@mkhatib I just tried amp-app-banner and I do not see a problem there either, tried it with both viewer and non-viewer (I forced the non-viewer version not to use the native install banner). Closing the banner definitely moves the cursor to the body and swipe next/prev starts reading from top of the body again. we should follow up with #4942 Chrome/Android seems to be doing the right thing. (I have seen TalkBack get confused in general before anyway, I wonder if #4942 was just a flake)

**Comment by mkhatib** [comment]
Thanks @aghassemi for all the testing, I'll try to do some more and see if I can repro what happened in that issue. From what I read about dialogs and similar UI components, re-setting the focus is a common accessibility issue. Will do some more testing and research about this.

**Comment by mkhatib** [comment]
Closing this as we couldn't repro the original issue filed.
